src/usage_metrics/resources/sqlite.py

```diff
@@ -49,26 +50,43 @@ def append_df_to_table(self, df: pd.DataFrame, table_name: str) -> None:
             table_name in usage_metrics_metadata.tables
         ), f"""{table_name} does not have a database schema defined.
             Create a schema one in usage_metrics.models."""
+        table_obj = usage_metrics_metadata.tables[table_name]
 
         if self.clobber:
-            table_obj = usage_metrics_metadata.tables[table_name]
             usage_metrics_metadata.drop_all(self.engine, tables=[table_obj])
 
-        # TODO: could also get the insert_ids already in the database
-        # and only append the new data.
+        # Get primary key column(s) of dataframe, and check against
+        # already-existing data.
+        pk_cols = [
+            pk_column.name for pk_column in table_obj.primary_key.columns.values()
+        ]
+        tbl = sa.Table(table_name, sa.MetaData(), autoload_with=self.engine)
+        query = sa.select(*[tbl.c[c] for c in pk_cols])  # Only select PK cols
+
         with self.engine.begin() as conn:
-            df.to_sql(
-                name=table_name,
-                con=conn,
-                if_exists="append",
-                index=False,
-            )
+            # Get existing primary keys
+            existing_pks = pd.read_sql(sql=query, con=conn)
+            i1 = df.set_index(pk_cols).index
+            i2 = existing_pks.set_index(pk_cols).index
+            # Only update primary keys that aren't in the database
+            df_new = df[~i1.isin(i2)]
+            if df_new.empty:
+                context.log.warn(
+                    "All records already loaded, not writing any data. Clobber the database if you want to overwrite this data."
+                )
+            else:
+                df_new.to_sql(
+                    name=table_name,
+                    con=conn,
+                    if_exists="append",
+                    index=False,
+                    dtype={c.name: c.type for c in table_obj.columns},
+                )
```

Comment:
How can we share this logic between the two IO managers so it is only specified in one place?

Reply:
I tried to generalize this using `SQLIOManager` - need to just test that it works on Postgres as expected, but let me know what you think of this strategy?

Reply:
Tested on the Cloud SQL database and all is working as expected.

Reply:
Looks great thank you!